src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Conversion.cs

```diff
@@ -1297,6 +1313,84 @@ private BoundExpression RewriteDecimalConversion(BoundConversion oldNode, CSharp
             }
         }
 
+        private BoundExpression RewriteStaticMethodGroupConversion(CSharpSyntaxNode syntax, BoundExpression operand, MethodSymbol targetMethod, NamedTypeSymbol delegateType)
+        {
+            Debug.Assert(delegateType.IsDelegateType());
+            Debug.Assert((object)targetMethod != null);
+
+            var cacheContainer = ObtainCacheContainer(delegateType, targetMethod);
+            var cacheField = cacheContainer.ObtainCacheField(_factory, delegateType, targetMethod);
+
+            var orgSyntax = _factory.Syntax;
+            _factory.Syntax = syntax;
+
+            var boundCacheField = _factory.Field(null, cacheField);
+            var boundDelegateCreation = new BoundDelegateCreationExpression(syntax, operand, targetMethod, isExtensionMethod: false, type: delegateType)
+            {
+                WasCompilerGenerated = true
+            };
+
+            var rewrittenNode = _factory.Coalesce(boundCacheField, _factory.AssignmentExpression(boundCacheField, boundDelegateCreation));
+
+            _factory.Syntax = orgSyntax;
+
+            return rewrittenNode;
+        }
+
+        private DelegateCacheContainer ObtainCacheContainer(NamedTypeSymbol delegateType, MethodSymbol targetMethod)
+        {
+            switch (ChooseDelegateCacheContainerKind(_factory.TopLevelMethod, delegateType, targetMethod))
+            {
+                case DelegateCacheContainerKind.ModuleScopedConcrete:
+                    return _factory.ModuleBuilderOpt.DelegateCacheManager.ObtainContainer(delegateType);
+                case DelegateCacheContainerKind.TypeScopedConcrete:
+                    return _factory.CompilationState.TypeScopedDelegateCacheContainer;
+                case DelegateCacheContainerKind.MethodScopedGeneric:
+                    return MethodScopedGenericDelegateCacheContainer;
+                default:
+                    throw ExceptionUtilities.Unreachable;
+            }
+        }
+
+        private static DelegateCacheContainerKind ChooseDelegateCacheContainerKind(MethodSymbol currentMethod, NamedTypeSymbol delegateType, MethodSymbol targetMethod)
+        {
+            // First, if delegateType and targetMethod are both concrete to teeth, consider a module scoped cache container, because
+            // a. targetMethod could be somewhere outside of currentMethod.ContainingType, there might be another method somewhere else want to use this cache
+            // b. currentMethod.ContainingType or/and its ancestors could be generic, making the container type scoped could possibly cause multiple instances at runtime
+            // There are cases when targetMethod or/and delegateType being private or only visible inside currentMethod.ContainingType by code, though
+            var fullyConcreteChecker = FullyConcreteChecker.Instance;
```

Comment:
We probably can skip all the visiting if we know that containing type !IsGeneric and the currentMethod is not generic.